homeassistant/components/mqtt/util.py

```diff
@@ -41,6 +53,56 @@ def mqtt_config_entry_enabled(hass: HomeAssistant) -> bool | None:
     return not bool(hass.config_entries.async_entries(DOMAIN)[0].disabled_by)
 
 
+async def async_wait_for_mqtt_client(hass: HomeAssistant) -> bool:
+    """Wait for the MQTT client to become available.
+
+    Waits when mqtt set up is in progress,
+    It is not needed that the client is connected.
+    Returns True if the mqtt client is available.
+    Returns False when the client is not available.
+    """
+    if not mqtt_config_entry_enabled(hass):
+        return False
+
+    entry = hass.config_entries.async_entries(DOMAIN)[0]
+    if entry.state == ConfigEntryState.LOADED:
+        return True
+
+    valid_states = {ConfigEntryState.SETUP_IN_PROGRESS, ConfigEntryState.LOADED}
+    state_reached_future: asyncio.Future[bool]
+    if DATA_MQTT_AVAILABLE not in hass.data:
+        hass.data[DATA_MQTT_AVAILABLE] = state_reached_future = asyncio.Future()
+    else:
+        state_reached_future = hass.data[DATA_MQTT_AVAILABLE]
+        if state_reached_future.done():
+            return state_reached_future.result()
+
+    @callback
+    def _async_entry_changed(
+        change: ConfigEntryChange, event_entry: ConfigEntry
+    ) -> None:
+        if event_entry is entry and change is ConfigEntryChange.UPDATED:
+            # SETUP_IN_PROGRESS: allow integration to set the result
+            if entry.state == ConfigEntryState.SETUP_IN_PROGRESS:
+                return
+            # Set the result based on the state
+            # False if when entered an error state else True
+            if not state_reached_future.done():
+                state_reached_future.set_result(entry.state in valid_states)
+
+    unsub = async_dispatcher_connect(
+        hass, SIGNAL_CONFIG_ENTRY_CHANGED, _async_entry_changed
+    )
+    try:
+        async with async_timeout.timeout(AVAILABILITY_TIMEOUT):
+            # Await the client setup or an error state was received
+            return await state_reached_future
+    except asyncio.TimeoutError:
+        return False
+    finally:
+        unsub()
```

Comment:
I think we should unrevert https://github.com/home-assistant/core/pull/82085

So we don't have mqtt knowing about `SIGNAL_CONFIG_ENTRY_CHANGED`

Reply:
Right, you are right. In fact I reïmplemented a parts of it.

Reply:
See #91706